### ISSUE TITLE
Rewrite `proxy-chain`

### ISSUE THREAD

**Issue by szmarczak** [apify/proxy-chain#152]
There are some problems with the current implementation:
- [`handler_base.js` always accesses the underlying socket](https://github.com/apify/proxy-chain/blob/0043ac9db2c04fffdd53fe7ee7d3a11183648555/src/handler_base.js#L67-L73), which is not necessary e.g. for normal HTTP requests
- tunnels can be simplified (current logic in general is complex imo)
- contains lots of legacy code such as [this](https://github.com/apify/proxy-chain/blob/0043ac9db2c04fffdd53fe7ee7d3a11183648555/src/handler_base.js#L41-L49)
- some utils can be removed make use of what is available in Node.js 14 e.g. `http.valdiateHeaderValue`

@jancurn What do you think?

**Comment by jancurn** [comment]
@szmarczak The proposed changes sound good, but altogether this would be quite a big change that wouldn't bring a direct benefit to users, and could potentially break some things. So perhaps it would make sense to do the changes sequentially rather than all at once, e.g. start with TypeScript, then remove the legacy code etc.? Also please discuss with @mnmkng the priority with respect to other projects

**Comment by mnmkng** [comment]
Since we would release the TS version as major, I think it would make sense to add other potentially breaking changes into that release. The question is, how long would the rewrite take if we were to do TS only and how long it would take if we were to do TS + all the upgrades. @szmarczak can you estimate it?

**Comment by B4nan** [comment]
I think it make sense to do all of those as part of a single major bump (we do major bumps for TS rewrites just to be safe, there can be build complications if people use TS, as well as to allow some minor/internal BCs during the rewrite).

**Comment by jancurn** [comment]
My thinking was it would be quite hard to review all these changes mixed into one giant PR. So ideally if we could review and ship the changes separately, even as alpha releases, it would make the review simpler.

Also, regarding "tunnels can be simplified", I'd be quite curious to see how.

**Comment by szmarczak** [comment]
> My thinking was it would be quite hard to review all these changes mixed into one giant PR.

I agree, however if we were to simplify the tunnels then the server api would change as well etc. in the end it would become a giant PR too.

I think I can do most of this in two or three days.

**Comment by mnmkng** [comment]
When we did the previous TS rewrites, we also did it step-by-step. I think it would make sense to split the PRs into at least the four above-listed topics.

**Comment by szmarczak** [comment]
Ok, I'll try to rewrite it step by step then. Might take more time.

**Comment by B4nan** [comment]
Giving it more time is fine, it will make reviewing much faster and feedback loop shorter.
